include/rocksdb/utilities/agg_merge.h

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <deque>
+#include <string>
+#include <vector>
+
+#include "rocksdb/merge_operator.h"
+#include "rocksdb/slice.h"
+
+namespace ROCKSDB_NAMESPACE {
+// A class used to aggregate data per key/value. The plug-in function is
+// implemented and registered using AddAggregator(). And then use it
+// with merge operator created using CreateAggMergeOperator().
+class Aggregator {
+ public:
+  virtual ~Aggregator() {}
+  // The input list is ordered with insertion order, with the latest one
+  // to be oldest. The oldest one might be from Get().
+  // Return whether aggregation succeeded. False for aggregation error.
+  virtual bool Aggregate(const std::vector<Slice>&,
+                         std::string* result) const = 0;
+
+  // True if a partial aggregation should be invoked. Some aggregators
+  // might opt to skip partial aggregation if possible.
+  virtual bool DoPartialAggregate() const { return true; }
+};
+
+// The function adds aggregation plugin by function name. It is used
+// by all the aggregation operator created using CreateAggMergeOperator().
+// It's currently not thread safe to run concurrently with the aggregation
+// merge operator. It is recommended that all the aggregation function
+// is added before calling CreateAggMergeOperator().
+Status AddAggregator(const std::string& function_name,
+                     std::unique_ptr<Aggregator>&& agg);
+
```

Comment:
I think about it a little bit and didn't figure out how this can be done. Lookup needs very good performance, as it needs to be looked up for every Get() and when processing each key/value in flush/compaction. It doesn't feel that ObjectRegistry is built for such a scenario. We can definitely re-evaluate it when the feature is turned into a native support of RocksDB. Right now, the feature lives outside RocksDB and there is no need to configure it using a string.

Reply:
From my reading of the code, the AddAggregator is done to configure an AggMergeOperator and not on every Get Operation.  Which is the use case the ObjectRegistry was built for.

Reply:
Maybe I didn't understand your proposal. If we use ObjectRegistry to register the functions, where do you suggest we retrieve those functions?